### PR TITLE
[FLINK-28943] Add Transformer and Estimator for MaxAbsScaler

### PR DESCRIPTION
**Pull request by weibozhao**
## What is the purpose of the change
Add Transformer and Estimator of MaxAbsScaler in FlinkML. 
## Brief change log
* Add MaxAbsScaler (Java & Python)
* Add MaxAbsScalerModel (Java & Python)
* Add unit test for MaxAbsScaler and MaxAbsScalerModel (Java & Python)
* Add example code for MaxAbsScaler and MaxAbsScalerModel (Java&Python)
## Does this pull request potentially affect one of the following parts:
Dependencies (does it add or upgrade a dependency): (no)
The public API, i.e., is any changed class annotated with @public(Evolving): (no)
Does this pull request introduce a new feature? (no)
If yes, how is the feature documented? (Java doc)

### CODE REVIEW

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                inputData
+                        .transform(
+                                "reduceInEachPartition",
+                                inputData.getType(),
+                                new MaxAbsReduceFunctionOperator())
+                        .transform(
+                                "reduceInFinalPartition",
+                                inputData.getType(),
+                                new MaxAbsReduceFunctionOperator())
+                        .setParallelism(1);
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:87]: Would it be better to reuse `DataStreamUtils.reduce` for this part?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        DataStream<MaxAbsScalerModelData> modelData =
+                DataStreamUtils.mapPartition(
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:89]: Given that the upstream operator would generate only one stream record, it seems unnecessary to use `DataStreamUtils.mapPartition` here. Maybe a simple `map` function is enough.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /**
+     * A stream operator to compute the max values in each partition of the input bounded data
+     * stream.
+     */
+    private static class MaxAbsReduceFunctionOperator extends AbstractStreamOperator<Vector>
+            implements OneInputStreamOperator<Vector, Vector>, BoundedOneInput {
+        private ListState<DenseVector> maxState;
+        private DenseVector maxVector;
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:113]: If all input records are sparse vectors, this practice might waste much memory space. Is there any better way to improve this implementation?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            maxState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "maxState", TypeInformation.of(DenseVector.class)));
+            OperatorStateUtils.getUniqueElement(maxState, "maxState").ifPresent(x -> maxVector = x);
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:178]: Some SuppressWarnings like this are redundant. Could you please check and remove them?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            this.inputCol = inputCol;
+        }
+
+        @Override
+        public Row map(Row row) {
+            if (scaleVector == null) {
+                MaxAbsScalerModelData maxAbsScalerModelData =
+                        (MaxAbsScalerModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+                scaleVector = maxAbsScalerModelData.maxVector;
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:146]: Could you please add document, support and tests for cases in which `maxVector` contains 0 values, which means `1.0 / scaleVector.values[i]` is invalid?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                for (int i = 0; i < scaleVector.size(); ++i) {
+                    if (scaleVector.values[i] != 0) {
+                        scaleVector.values[i] = 1.0 / Math.abs(scaleVector.values[i]);
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:149]: Spark's implementation is similar to `1.0 / scaleVector.values[i]`, instead of `1.0 / Math.abs(scaleVector.values[i])`. Would it be better to follow Spark's practice?

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.test.util.TestBaseUtils.compareResultCollections;
+import static org.junit.Assert.assertEquals;
+
+/** Tests {@link MaxAbsScaler} and {@link MaxAbsScalerModel}. */
+public class MaxAbsScalerTest {
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:55]: Could you please add a test to verify the case when all values of input vectors are negative values? There might be a bug here?

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        assertEquals(
+                Collections.singletonList("maxVector"),
+                modelData.getResolvedSchema().getColumnNames());
+        DataStream<Row> output = tEnv.toDataStream(modelData);
+        List<Row> modelRows = IteratorUtils.toList(output.executeAndCollect());
+        assertEquals(new DenseVector(new double[] {200.0, 400.0}), modelRows.get(0).getField(0));
+    }
+
+    @Test
+    public void testSetModelData() throws Exception {
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:262]: Could you please add a test case that sets a model data that contains zero and negative values, and make sure the resulting behavior is the same as that in Spark?

**Review by yunfengzhou-hub — COMMENTED** [review]
Thanks for the PR. Left some comments as below.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java (hunk 5 above)
Comment (weibozhao) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:146]: If maxVector contains zero values, that means all values in these features are zero. I have added a ut for this case.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java (hunk 7 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:55]: I have fixed this bug.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java (hunk 1 above)
Comment (weibozhao) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:87]: I think `DataStreamUtils.reduce` is not suit here. When the input table only has one vector and this vector has some negative values, the output data will not be correct.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java (hunk 3 above)
Comment (weibozhao) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:113]: I think this case will not happen.  If all input records are sparse vectors, the maxAbsVector need not be a sparse vector. If a lot of values in maxAbsVector are zeros, which means the input data has a lot of invalid features, we need to check the input data.

Comment (yunfengzhou-hub) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java]: Got it. I'm OK with the current implementation.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+        @Override
+        public Row map(Row row) {
+            if (scaleVector == null) {
+                MaxAbsScalerModelData maxAbsScalerModelData =
+                        (MaxAbsScalerModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+                scaleVector = maxAbsScalerModelData.maxVector;
+                for (int i = 0; i < scaleVector.size(); ++i) {
+                    if (scaleVector.values[i] != 0) {
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:148]: Spark would set the scale to 1 if a value in maxVector is 0. Would it be better to follow the existing practice in Spark?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Row> data = tEnv.toDataStream(inputs[0]);
+        DataStream<MaxAbsScalerModelData> maxAbsScalerModel =
+                MaxAbsScalerModelData.getModelDataStream(modelDataTable);
+        final String broadcastModelKey = "broadcastModelKey";
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(
+                                inputTypeInfo.getFieldTypes(), TypeInformation.of(Vector.class)),
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:84]: nit: VectorTypeInfo.INSTANCE

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Model which do a MaxAbsScaler operation using the model data computed by {@link MaxAbsScaler}.
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:48]: nit: "does". Same for `MinMaxScaler` and their python documents.

Or we can refer to the JavaDoc of `StandardScalerModel` and keep all the three scaler algorithms with the same JavaDoc template.

**Review by yunfengzhou-hub — COMMENTED** [review]
Thanks for the update. Left some comments as below.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+                    }
+                } else if (currentValue instanceof SparseVector) {
+                    int[] indices = ((SparseVector) currentValue).indices;
+                    double[] values = ((SparseVector) currentValue).values;
+                    for (int i = 0; i < indices.length; ++i) {
+                        maxAbsVector.values[indices[i]] =
+                                Math.max(maxAbsVector.values[indices[i]], Math.abs(values[i]));
+                    }
+                } else {
+                    throw new RuntimeException(
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:152]: nit: Let's throw a more detailed exception here, like `IllegalArgumentException`.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            maxAbsState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "maxState", TypeInformation.of(DenseVector.class)));
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:165]: nit: `DenseVectorTypeInfo`.

**Review by yunfengzhou-hub — COMMENTED** [review]
Thanks for the update. Left two small comments. The rest of this PR LGTM.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        public void processElement(StreamRecord<Vector> streamRecord) {
+            Vector currentValue = streamRecord.getValue();
+            if (currentValue == null) {
+                throw new RuntimeException("Input column data cannot be null.");
+            }
+
+            if (maxAbsVector == null) {
+                int vecSize = currentValue.size();
+                maxAbsVector = new DenseVector(vecSize);
+                if (currentValue instanceof DenseVector) {
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:125]: How about we simplify the code as:
```
maxAbsVector = maxAbsVector == null? new DenseVector(currentValue.size): maxAbsVector
```
and remove the code from Line#121 to Line#139?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+                } else if (currentValue instanceof SparseVector) {
+                    int[] indices = ((SparseVector) currentValue).indices;
+                    double[] values = ((SparseVector) currentValue).values;
+
+                    for (int i = 0; i < indices.length; ++i) {
+                        maxAbsVector.values[indices[i]] =
+                                Math.max(maxAbsVector.values[indices[i]], Math.abs(values[i]));
+                    }
+                } else {
+                    throw new IllegalArgumentException(
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:159]: nit: This line could be removed because the Vector could only be SparseVector or DenseVector for now.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            maxAbsState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "maxState", DenseVectorTypeInfo.INSTANCE));
+
+            OperatorStateUtils.getUniqueElement(maxAbsState, "maxState")
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:174]: nit: maxAbsState

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            maxAbsState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "maxState", DenseVectorTypeInfo.INSTANCE));
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:172]: nit: maxAbsState

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+            throws IOException {
+        MaxAbsScalerModel model = ReadWriteUtils.loadStageParam(path);
+
+        Table modelDataTable =
+                ReadWriteUtils.loadModelData(
+                        tEnv, path, new MaxAbsScalerModelData.ModelDataDecoder());
+        return model.setModelData(modelDataTable);
+    }
+
+    /** This operator loads model data and predicts result. */
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:131]: nit: This operator --> This function.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+                for (int i = 0; i < scaleVector.size(); ++i) {
+                    if (scaleVector.values[i] != 0) {
+                        scaleVector.values[i] = 1.0 / scaleVector.values[i];
+                    } else {
+                        scaleVector.values[i] = 1.0;
+                    }
+                }
+            }
+
+            if (row.getFieldAs(inputCol) == null) {
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScalerModel.java:159]: Shall we remove Line#159 to Line#161 and assume that the `inputCol` field of the row is never null? We need to make sure that the behavior of the existing implementations are consistent.

Please refer to MinMaxScalerModel, StringIndexerModel and other existing algorithms.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) output).getTableEnvironment();
+
+        DataStream<Vector> stream =
+                tEnv.toDataStream(output)
+                        .map((MapFunction<Row, Vector>) row -> row.getFieldAs(outputCol));
+
+        List<Vector> result = IteratorUtils.toList(stream.executeAndCollect());
+        if (isSparse) {
+            compareResultCollections(
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:144]: Shall we make expectedData an argument of `verifyPredictionResult` function?

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+                                Row.of(Vectors.dense(200, 400))));
+
+        Table trainDataWithInvalidData =
+                tEnv.fromDataStream(env.fromCollection(trainData)).as("input");
+
+        try {
+            MaxAbsScaler maxAbsScaler = new MaxAbsScaler();
+            MaxAbsScalerModel maxAbsScalerModel = maxAbsScaler.fit(trainDataWithInvalidData);
+            IteratorUtils.toList(
+                    tEnv.toDataStream(maxAbsScalerModel.getModelData()[0]).executeAndCollect());
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:200]: Shall we add `fail()` next line? For now, the test passes even we do not get into the exception.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+        } catch (Exception e) {
+            assertEquals(
+                    "Input column data cannot be null.",
+                    ExceptionUtils.getRootCause(e).getMessage());
+        }
+    }
+
+    @Test
+    public void testTransformDataWithNullValue() throws Exception {
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:209]: Probably we should not allow transforming invalid data.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+
+                    for (int i = 0; i < indices.length; ++i) {
+                        maxAbsVector.values[indices[i]] =
+                                Math.max(maxAbsVector.values[indices[i]], Math.abs(values[i]));
+                    }
+                }
+            } else {
+                Preconditions.checkArgument(
+                        currentValue.size() == maxAbsVector.size(),
+                        "CurrentValue should has same size with maxVector.");
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java:142]: nit: How about update the error message as: `The training data should all have same dimensions`.

**Review by zhipeng93 — COMMENTED** [review]
Thanks for the PR. Left some comments below.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+                                Row.of((Object) null),
+                                Row.of(Vectors.dense(6.1, 8.1)),
+                                Row.of(Vectors.dense(200, 400))));
+
+        Table trainDataWithInvalidData =
+                tEnv.fromDataStream(env.fromCollection(trainData)).as("input");
+
+        try {
+            MaxAbsScaler maxAbsScaler = new MaxAbsScaler();
+            maxAbsScaler.fit(trainDataWithInvalidData);
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/MaxAbsScalerTest.java:193]: Could you insert a `fail();` after this line to verify that the code never goes there?

**Review by zhipeng93 — COMMENTED** [review]
Thanks for the PR. LGTM overall except one minor comment.

**Comment by zhipeng93** [comment]
LGTM.
